README.md
### NOTE: This release branch is no longer maintained and is deprecated. Please consider using the `release-latest` branch for the latest stable sources.


# h264_video_encoder


## Overview
This package provides a ROS Node that will encode a stream of images into an H264 video stream.

**Keywords**: ROS, AWS, Kinesis

### License
The source code is released under [LGPL 2.1]. However, this package uses `h264_encoder_core` which incorporates several different encoding components which may further restrict the license. By default, x264 is used for software encoding, thereby applying GPL to all of h264_video_encoder.

**Author**: AWS RoboMaker<br/>
**Affiliation**: [Amazon Web Services (AWS)]<br/>
**Maintainer**: AWS RoboMaker, ros-contributions@amazon.com

### Supported ROS Distributions
- Kinetic
- Lunar
- Melodic

### Build status
* Travis CI: [![Build Status](https://travis-ci.org/aws-robotics/kinesisvideo-encoder-ros1.svg?branch=master)](https://travis-ci.org/aws-robotics/kinesisvideo-encoder-ros1)
 * ROS build farm:
   * v1.0.0:
     * ROS Kinetic @ u16.04 Xenial [![Build Status](http://build.ros.org/job/Kbin_uX64__h264_video_encoder__ubuntu_xenial_amd64__binary/badge/icon)](http://build.ros.org/job/Kbin_uX64__h264_video_encoder__ubuntu_xenial_amd64__binary)


## Installation

### Building from Source
Create a ROS workspace and a source directory

    mkdir -p ~/ros-workspace/src

To build from source, clone the latest version from master branch and compile the package.

- Clone the package into the source directory

        cd ~/ros-workspace/src
        git clone https://github.com/aws/aws-ros-utils-common.git
        git clone https://github.com/aws/aws-ros-utils-ros1.git
        git clone https://github.com/aws/aws-ros-kinesisvideo-encoder-common.git
        git clone https://github.com/aws/aws-ros-kinesisvideo-encoder-ros1.git

- Install dependencies

        cd ~/ros-workspace && sudo apt-get update
        rosdep install --from-paths src --ignore-src -r -y

- Build the packages

        cd ~/ros-workspace && colcon build

- Configure ROS library Path

        source ~/ros-workspace/install/setup.bash

- Build and run the unit tests

        colcon build --packages-select h264_video_encoder --cmake-target tests
        colcon test --packages-select h264_video_encoder h264_encoder core && colcon test-results --all


### Building on Cloud9 - Cross Compilation

- In RoboMaker's Cloud9, start with an empty workspace and in the Cloud9 console:

       # build docker image
       cd /opt/robomaker/cross-compilation-dockerfile/
       sudo bin/build_image.bash  # this step will take a while

       # create workspace
       mkdir -p ~/environment/robot_ws/src
       cd ~/environment/robot_ws/src
       git clone https://github.com/aws-robotics/kinesisvideo-encoder-common.git
       git clone https://github.com/aws-robotics/kinesisvideo-encoder-ros1.git

       # run docker image
       cd ..
       sudo docker run -v $(pwd):/ws -it ros-cross-compile:armhf

- Now you're inside the cross-compilation docker container

       # build the workspace
       cd ws
       apt update
       rosdep install --from-paths src --ignore-src -r -y  # this step will take a while
       colcon build --build-base armhf_build --install-base armhf_install
       colcon bundle --build-base armhf_build --install-base armhf_install --bundle-base armhf_bundle --apt-sources-list /opt/cross/apt-sources.yaml   # this step will take a while
       exit

- Now you're oustide the cross-compilation docker container

       # for more on copying s3 buckets see: https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html
       aws s3 cp armhf_bundle/output.tar.gz s3://<bucket_name_in_your_robomaker_account>/h264_video_encoder.armhf.tar

## Launch Files
A launch file called `h264_video_encoder.launch` is included in this package. The launch file uses the following arguments:

| Arg Name | Description |
| --------- | ------------ |
| node_name | (optional) The name the H264 encoder node should be launched with. If not provided, the node name will default to `h264_video_encoder` |
| config_file | (optional) A path to a rosparam config file. |

An example launch file called `sample_application.launch` is included in this project that gives an example of how you can include this node in your project and provide it with arguments.


## Usage

### Running the node
To launch the H264 encoder node, you can run the following command:

    roslaunch h264_video_encoder sample_application.launch


## Configuration File and Parameters
An example configuration file called `sample_configuration.yaml` is provided for running the H264 encoder node on a Raspberry Pi based system.
When the parameters are absent in the ROS parameter server, default values are used, thus all parameters are optional. See table below for details.

| Parameter Name | Description | Type |
| ------------- | -----------------------------------------------------------| ------------- |
| queue_size | (optional) The maximum number of incoming and outgoing messages to be queued towards the subscribed and publishing topics. | integer |
| output_width | (optional) The desired width (in pixels) of each frame in the encoded video output. | integer |
| output_height | (optional) The desired height (in pixels) of each frame in the encoded video output. | integer |
| fps_numerator | (optional) The desired frames per second (the numerator portion when expressing FPS as a rational number) for the encoded video output. | integer |
| fps_denominator | (optional) The desired frames per second (the denominator portion when expressing FPS as a rational number) for the encoded video output. | integer |
| bitrate | (optional) The desired bitrate (in bits per second) of the encoded video output. | integer |


## Node Details

#### Published Topics
| Topic Name | Message Type | Description |
| ---------- | ------------ | ----------- |
| *Configurable* (default="video/encoded") | kinesis_video_msgs/KinesisVideoFrame | The node will publish to a topic of a given name. Each message being published contains a chunk of the video stream, usually per video frame. |

#### Subscribed Topics
| Topic Name | Message Type | Description |
| ---------- | ------------ | ----------- |
| *Configurable* (default="/raspicam_node/image") | sensor_msgs/Image | The node will subscribe to a topic of a given name. The data is expected to be a stream of images from a source (such as a Raspberry Pi camera). |


## Bugs & Feature Requests
Please contact the team directly if you would like to request a feature.

Please report bugs in [Issue Tracker].


[Amazon Web Services (AWS)]: https://aws.amazon.com/
[LGPL 2.1]: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
[Issue Tracker]: https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues
[ROS]: http://www.ros.org


h264_video_encoder/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package h264_video_encoder
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.1.4 (2019-08-29)
------------------
* increment patch number for bloom release (`#34 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/34>`_)
* Adds h264_encoder_core as exec_depend (`#33 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/33>`_)
  This commit adds h264_encoder_core as an execution dependency. It also puts the dependencies into alphabetic order.
* increment patch version (`#31 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/31>`_)
  Signed-off-by: Miaofei <miaofei@amazon.com>
* Use standard CMake macros for adding gtest/gmock tests (`#25 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/25>`_)
  * modify h264_video_encoder to use add_rostest_gmock()
  Signed-off-by: Miaofei <miaofei@amazon.com>
  * update travis.yml to be compatible with specifying multiple package names
  Signed-off-by: Miaofei <miaofei@amazon.com>
* Update package.xml for 1.1.2 release
  Signed-off-by: Ryan Newell <ryanewel@amazon.com>
* fix issue with uninitialized frame number (`#19 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/19>`_)
  Signed-off-by: Miaofei <miaofei@amazon.com>
* Add gtest and gmock as test dependencies
* Update package.xml
* Merge pull request `#8 <https://github.com/aws-robotics/kinesisvideo-encoder-ros1/issues/8>`_ from ryanewel/master
  increases unit test code coverage
* increases unit test code coverage
* Contributors: AAlon, Abby Xu, M. M, Nick Burek, Ross Desmond, Ryan Newell, ryanewel


h264_video_encoder/test/h264_video_encoder_test.cpp
/*
 *  Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <aws_ros1_common/sdk_utils/ros1_node_parameter_reader.h>
#include <aws/core/utils/logging/AWSLogging.h>
#include <aws/core/utils/logging/LogMacros.h>
#include <gtest/gtest.h>
#include <h264_encoder_core/h264_encoder.h>
#include <h264_encoder_core/h264_encoder_node_config.h>
#include <image_transport/image_transport.h>
#include <kinesis_video_msgs/KinesisImageMetadata.h>
#include <kinesis_video_msgs/KinesisVideoFrame.h>
#include <ros/ros.h>
#include <sensor_msgs/image_encodings.h>

using namespace Aws::Utils::Encoding;
using namespace Aws::Utils::Logging;

namespace Aws {
namespace Kinesis {

void InitializeEncoder(const sensor_msgs::ImageConstPtr & msg,
                       std::unique_ptr<H264Encoder> & encoder,
                       const Aws::Client::ParameterReaderInterface & param_reader);

void ImageCallback(const sensor_msgs::ImageConstPtr & msg, const H264Encoder * encoder,
                   uint64_t & frame_num, kinesis_video_msgs::KinesisImageMetadata & metadata,
                   ros::Publisher & pub);

void InitializeCommunication(ros::NodeHandle & nh,
                             ros::Subscriber& metadata_sub,
                             image_transport::Subscriber& image_sub,
                             ros::Publisher& pub,
                             std::unique_ptr<H264Encoder>& encoder,
                             uint64_t & frame_num,
                             kinesis_video_msgs::KinesisImageMetadata & metadata,
                             Aws::Client::Ros1NodeParameterReader & param_reader
                            );

}  // namespace Kinesis
}  // namespace Aws

constexpr static char kDefaultPublicationTopicName[] = "/video/encoded";
constexpr static char kDefaultSubscriptionTopicName[] = "/raspicam_node/image";
constexpr static char kDefaultMetadataTopicName[] = "/image_metadata";
constexpr static int kDefaultWidth = 410;
constexpr static int kDefaultHeight = 308;
constexpr static int kBytesPerPixel = 3;  // 3 color channels (red, green, blue)
constexpr int kNumTestFrames = 30;


class H264EncoderNodeSuite : public ::testing::Test
{
public:

  constexpr static const std::string & kDefaultEncoding = sensor_msgs::image_encodings::RGB8;

  H264EncoderNodeSuite() {}

protected:
  void SetUp() override
  {
    ros::Time::init();

    default_msg = boost::make_shared<sensor_msgs::Image>();
    default_msg->header.seq = 0;
    default_msg->header.frame_id = "";
    default_msg->header.stamp = ros::Time::now();
    default_msg->height = kDefaultHeight;
    default_msg->width = kDefaultWidth;
    default_msg->encoding = kDefaultEncoding;
    default_msg->step = kBytesPerPixel * kDefaultWidth;
  }

  void KinesisVideoCallback(const kinesis_video_msgs::KinesisVideoFrame::ConstPtr & frame)
  {
    if (frame->index > 0) {
      EXPECT_GT(frame->index, prev_frame.index);
      EXPECT_EQ(frame->duration, prev_frame.duration);
      EXPECT_EQ((frame->decoding_ts - prev_frame.decoding_ts) % frame->duration, 0);
      EXPECT_EQ((frame->presentation_ts - prev_frame.presentation_ts) % frame->duration, 0);
      EXPECT_GT(frame->presentation_ts, prev_frame.presentation_ts);
    }

    prev_frame = *frame;
    if (0 == prev_frame.index) {
      fwrite(prev_frame.codec_private_data.data(), 1, prev_frame.codec_private_data.size(),
             debug_file);
    }
    fwrite(prev_frame.frame_data.data(), 1, prev_frame.frame_data.size(), debug_file);
  }

  sensor_msgs::ImagePtr default_msg;
  std::unique_ptr<H264Encoder> encoder;
  Aws::Client::Ros1NodeParameterReader param_reader;

  kinesis_video_msgs::KinesisVideoFrame prev_frame;
  FILE * debug_file;
};

/**
 * Tests the callback of the H264 Encoder Node that performs the initialization
 */
TEST_F(H264EncoderNodeSuite, EncoderInit)
{
  std::unique_ptr<H264Encoder> invalid_encoder;
  sensor_msgs::ImagePtr invalid_msg = boost::make_shared<sensor_msgs::Image>();
  invalid_msg->encoding = "InvalidEncoding";
  Aws::Kinesis::InitializeEncoder(invalid_msg, invalid_encoder, param_reader);
  EXPECT_EQ(invalid_encoder, nullptr);
  Aws::Kinesis::InitializeEncoder(default_msg, encoder, param_reader);
  EXPECT_NE(encoder, nullptr);
}

static void RainbowColor(const float h, uint8_t & r_out, uint8_t & g_out, uint8_t & b_out)
{
  int i = 6.0f * h;
  float f = 6.0f * h - i;
  float t = f;
  float q = 1.0f - f;

  float r, g, b;
  switch (i % 6) {
    case 0:
      r = 1.0f;
      g = t;
      b = 0.0f;
      break;
    case 1:
      r = q;
      g = 1.0f;
      b = 0.0f;
      break;
    case 2:
      r = 0.0f;
      g = 1.0f;
      b = t;
      break;
    case 3:
      r = 0.0f;
      g = q;
      b = 1.0f;
      break;
    case 4:
      r = t;
      g = 0.0f;
      b = 1.0f;
      break;
    case 5:
      r = 1.0f;
      g = 0.0f;
      b = q;
      break;
  }

  r_out = std::lround(255.0f * r);
  g_out = std::lround(255.0f * g);
  b_out = std::lround(255.0f * b);
}

void CreateImageMsg(sensor_msgs::ImagePtr & msg, int frame_num)
{
    ++msg->header.seq;
    msg->header.stamp = ros::Time::now();

    // prepare a dummy image
    int shift = static_cast<float>(frame_num) / (kNumTestFrames - 1) * kDefaultWidth;
    for (int y = 0; y < kDefaultHeight; ++y) {
      for (int x = 0; x < kDefaultWidth; ++x) {
        uint8_t r, g, b;
        RainbowColor(static_cast<float>((x + shift) % kDefaultWidth) / kDefaultWidth, r, g, b);
        msg->data[kBytesPerPixel * y * kDefaultWidth + kBytesPerPixel * x + 0] = r;
        msg->data[kBytesPerPixel * y * kDefaultWidth + kBytesPerPixel * x + 1] = g;
        msg->data[kBytesPerPixel * y * kDefaultWidth + kBytesPerPixel * x + 2] = b;
      }
    }
}

/**
 * Tests the callback of the H264 Encoder Node that performs the encoding
 */
TEST_F(H264EncoderNodeSuite, EncoderCallback)
{
  Aws::Kinesis::InitializeEncoder(default_msg, encoder, param_reader);
  EXPECT_NE(encoder, nullptr);

  ros::NodeHandle pub_node;
  ros::Publisher pub =
    pub_node.advertise<kinesis_video_msgs::KinesisVideoFrame>(kDefaultPublicationTopicName, 100);

  ros::NodeHandle sub_node;
  boost::function<void(const kinesis_video_msgs::KinesisVideoFrame::ConstPtr &)> callback;
  callback = [this](const kinesis_video_msgs::KinesisVideoFrame::ConstPtr & frame) -> void {
    this->KinesisVideoCallback(frame);
  };
  ros::Subscriber sub = sub_node.subscribe(kDefaultPublicationTopicName, 100, callback);

  default_msg->data.resize(kBytesPerPixel * kDefaultWidth * kDefaultHeight);
  debug_file = fopen("frames_encodercallback.bin", "wb");

  // let's encode 30 frames
  uint64_t prev_frame_index = 0, frame_index = 0;
  kinesis_video_msgs::KinesisImageMetadata metadata;
  for (int i = 0; i < kNumTestFrames; ++i) {
    CreateImageMsg(default_msg, i);
    Aws::Kinesis::ImageCallback(default_msg, encoder.get(), frame_index, metadata, pub);
    ros::spinOnce();

    EXPECT_GE(frame_index, prev_frame_index);
    prev_frame_index = frame_index;
  }

  fclose(debug_file);
  // you can dump the debug frames by executing: ffmpeg -i frames.bin -frames:v 10 -f image2
  // frame%03d.png
}


/**
 * Tests that InitializeCommunicaiton sets up the correct subscribers and publisher
 */
 
TEST_F(H264EncoderNodeSuite, InitializeCommunicaiton)
{
  ros::NodeHandle nh("~");
  ros::Publisher pub;
  image_transport::Subscriber image_sub;
  ros::Subscriber metadata_sub;
  uint64_t frame_num;
  kinesis_video_msgs::KinesisImageMetadata metadata;
  Aws::Client::Ros1NodeParameterReader param_reader;
  Aws::Kinesis::  InitializeCommunication(nh, metadata_sub, image_sub, pub,
                          encoder, frame_num, metadata, param_reader);
  
  EXPECT_EQ(kDefaultPublicationTopicName, pub.getTopic());
  EXPECT_EQ(kDefaultSubscriptionTopicName, image_sub.getTopic());
  EXPECT_EQ(kDefaultMetadataTopicName, metadata_sub.getTopic());

  
  debug_file = fopen("frames_intialize_communication.bin", "wb");
  // Test that callback function is properly set up
  // let's encode 30 frames
  ros::NodeHandle sub_node;
  boost::function<void(const kinesis_video_msgs::KinesisVideoFrame::ConstPtr &)> callback;
  callback = [this](const kinesis_video_msgs::KinesisVideoFrame::ConstPtr & frame) -> void {
    this->KinesisVideoCallback(frame);
  };
  ros::Subscriber sub = sub_node.subscribe(kDefaultPublicationTopicName, 100, callback);

  default_msg->data.resize(kBytesPerPixel * kDefaultWidth * kDefaultHeight);
  ros::NodeHandle pub_node("~");
  ros::Publisher image_pub =
    pub_node.advertise<sensor_msgs::Image>(kDefaultSubscriptionTopicName, 100);
  constexpr int kNumTestFrames = 30;
  uint64_t prev_frame_index = 0, frame_index = 0;
  for (int i = 0; i < kNumTestFrames; ++i) {
    CreateImageMsg(default_msg, i);
    image_pub.publish(default_msg);
    ros::spinOnce();
    EXPECT_GE(frame_index, prev_frame_index);
    prev_frame_index = frame_index;
  }
    fclose(debug_file);


}


int main(int argc, char ** argv)
{

  testing::InitGoogleTest(&argc, argv);
  ros::init(argc, argv, "test_h264_video_encoder");
  return RUN_ALL_TESTS();
}


h264_video_encoder/src/h264_video_encoder.cpp
/*
 *  Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <aws/core/utils/logging/AWSLogging.h>
#include <aws/core/utils/logging/LogMacros.h>
#include <aws_ros1_common/sdk_utils/logging/aws_ros_logger.h>
#include <aws_ros1_common/sdk_utils/ros1_node_parameter_reader.h>
#include <h264_encoder_core/h264_encoder.h>
#include <h264_encoder_core/h264_encoder_node_config.h>
#include <image_transport/image_transport.h>
#include <kinesis_video_msgs/KinesisImageMetadata.h>
#include <kinesis_video_msgs/KinesisVideoFrame.h>
#include <ros/ros.h>
#include <sensor_msgs/image_encodings.h>

#include <map>
#include <string>
#include <vector>

using namespace Aws::Utils::Encoding;
using namespace Aws::Utils::Logging;

namespace Aws {
namespace Kinesis {

const std::map<std::string, AVPixelFormat> SNSR_IMG_ENC_to_LIBAV_PIXEL_FRMT = {
  {sensor_msgs::image_encodings::RGB8, AV_PIX_FMT_RGB24},
  {sensor_msgs::image_encodings::BGR8, AV_PIX_FMT_BGR24},
  {sensor_msgs::image_encodings::RGBA8, AV_PIX_FMT_RGBA},
  {sensor_msgs::image_encodings::BGRA8, AV_PIX_FMT_BGRA}};

/**
 * Initialize the H264Encoder
 * @param msg the message from the image sensor through image transport
 * @param encoder reference to pointer that owns the H264Encoder instance. if
 *  the pointer is null, it will be modified to the address of the new H264Encoder instance
 * @param param_reader parameter reader used for reading the desired configuration of the encoder
 * output
 */
void InitializeEncoder(const sensor_msgs::ImageConstPtr & msg,
                       std::unique_ptr<H264Encoder> & encoder,
                       const Aws::Client::ParameterReaderInterface & param_reader)
{
  auto encoding_iter = SNSR_IMG_ENC_to_LIBAV_PIXEL_FRMT.find(msg->encoding);
  if (encoding_iter == SNSR_IMG_ENC_to_LIBAV_PIXEL_FRMT.end()) {
    AWS_LOGSTREAM_ERROR(__func__,
                        "Trying to work with unsupported encoding " << msg->encoding << "!");
    return;
  }

  encoder = std::unique_ptr<H264Encoder>(new H264Encoder());
  if (nullptr != encoder) {
    encoder->Initialize(msg->width, msg->height, encoding_iter->second, param_reader);
  }
}

void ImageCallback(const sensor_msgs::ImageConstPtr & msg, const H264Encoder * encoder,
                   uint64_t & frame_num, kinesis_video_msgs::KinesisImageMetadata & metadata,
                   ros::Publisher & pub)
{
  thread_local H264EncoderResult encoder_output;

  AwsError retcode = encoder->Encode(msg->data.data(), encoder_output);
  if (retcode != AWS_ERR_OK) {
    if (retcode == AWS_ERR_NULL_PARAM) {
      AWS_LOG_ERROR(__func__, "Encoder received empty data!");
    } else if (retcode == AWS_ERR_FAILURE) {
      AWS_LOG_ERROR(__func__, "Unknown encoding error occurred");
    } else if (retcode == AWS_ERR_EMPTY) {
      AWS_LOG_WARN(__func__, "Encoder returned empty frame");
    }
    return;
  }

  kinesis_video_msgs::KinesisVideoFrame frame;
  frame.index = frame_num;
  frame.flags = (encoder_output.key_frame) ? kKeyFrameFlag : kBPFrameFlag;
  frame.decoding_ts = encoder_output.frame_dts;
  frame.presentation_ts = encoder_output.frame_pts;
  frame.duration = encoder_output.frame_duration / 2;  // duration recommended to be set shorter
  frame.codec_private_data = encoder->GetExtraData();
  frame.frame_data = encoder_output.frame_data;
  frame.metadata.swap(metadata.metadata);

  pub.publish(frame);

  constexpr int kDbgMsgThrottlePeriod = 10;  // 10 seconds throttling period
  ROS_DEBUG_THROTTLE(kDbgMsgThrottlePeriod, "Published Frame #%lu (timestamp: %lu)\n", frame_num,
                     encoder_output.frame_pts);

  ++frame_num;
}

void InitializeCommunication(ros::NodeHandle & nh,
                             ros::Subscriber& metadata_sub,
                             image_transport::Subscriber& image_sub,
                             ros::Publisher& pub,
                             std::unique_ptr<H264Encoder>& encoder,
                             uint64_t & frame_num,
                             kinesis_video_msgs::KinesisImageMetadata & metadata,
                             Aws::Client::Ros1NodeParameterReader & param_reader
                            )
{
  //
  // reading parameters
  //
  H264EncoderNodeParams params;
  GetH264EncoderNodeParams(param_reader, params);


  pub = nh.advertise<kinesis_video_msgs::KinesisVideoFrame>(params.publication_topic,
                                                                           params.queue_size);
  //
  // subscribing to topic with callback
  //


  boost::function<void(const sensor_msgs::ImageConstPtr &)> image_callback;
  image_callback = [&](const sensor_msgs::ImageConstPtr & msg) -> void {
    if (0 < pub.getNumSubscribers()) {
      if (nullptr == encoder) {
        InitializeEncoder(msg, encoder, param_reader);
      }
      if (nullptr != encoder) {
        ImageCallback(msg, encoder.get(), frame_num, metadata, pub);
      }
    } else {
      frame_num = 0;
    }
  };

  image_transport::ImageTransport it(nh);
  image_sub =
    it.subscribe(params.subscription_topic, params.queue_size, image_callback);
  AWS_LOGSTREAM_INFO(__func__, "subscribed to " << params.subscription_topic << "...");

  boost::function<void(const kinesis_video_msgs::KinesisImageMetadata::ConstPtr &)>
    metadata_callback;
  metadata_callback = [&](const kinesis_video_msgs::KinesisImageMetadata::ConstPtr & msg) -> void {
    metadata.metadata.insert(metadata.metadata.end(), msg->metadata.begin(), msg->metadata.end());
  };
  metadata_sub =
    nh.subscribe(params.metadata_topic, params.queue_size, metadata_callback);
  AWS_LOGSTREAM_INFO(__func__, "subscribed to " << params.metadata_topic << " for metadata...");
}

AwsError RunEncoderNode(int argc, char ** argv)
{
  ros::init(argc, argv, "h264_video_encoder");
  ros::NodeHandle nh("~");
  
  Aws::Utils::Logging::InitializeAWSLogging(
    Aws::MakeShared<Aws::Utils::Logging::AWSROSLogger>("h264_video_encoder"));
  AWS_LOG_INFO(__func__, "Starting H264 Video Node...");

  ros::Publisher pub;
  image_transport::Subscriber image_sub;
  ros::Subscriber metadata_sub;
  std::unique_ptr<H264Encoder> encoder;
  uint64_t frame_num;
  kinesis_video_msgs::KinesisImageMetadata metadata;
  Aws::Client::Ros1NodeParameterReader param_reader;

  InitializeCommunication(nh, metadata_sub, image_sub, pub,
                          encoder, frame_num, metadata, param_reader);
  
  //
  // run the node
  //
  ros::spin();
  AWS_LOG_INFO(__func__, "Shutting down H264 Video Node...");
  Aws::Utils::Logging::ShutdownAWSLogging();
  return AWS_ERR_OK;
}

}  // namespace Kinesis
}  // namespace Aws
